### PR TITLE
App framework - Post menu and channel info bindings, App forms

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR adds functionality for the App framework, specifically:

- Post menu bindings
- Channel info bindings
- Fixes for App forms

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-43210
https://mattermost.atlassian.net/browse/MM-43211
https://mattermost.atlassian.net/browse/MM-43209

#### Known remaining issues

- [x] App form submit button is not pressable for some reason - FIXED
- [x] Icons need to support HTTP URLs - DONE
- [x] Post menu items scrolling doesn't work - FIXED
- [x] Radio button styling for dark theme (see below) - [Follow up ticket](https://mattermost.atlassian.net/browse/MM-48732)

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

Android Simulator - Pixel 5

#### Release Note

```release-note
Support for the App framework has been added
```

### CODE REVIEW

File: app/screens/apps_form/apps_form_component.tsx [hunk 1]
```diff
@@ -369,7 +364,11 @@ function AppsFormComponent({
                 return [];
             }
         }
-    }, []);
+    }, [form, values, performLookupCall, setErrors, intl]);
+
+    // for some reason pressing the button does not run the callback, so temporarily setting the close button to this callback
+    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, handleSubmit, [handleSubmit]);
+    useNavButtonPressed(SUBMIT_BUTTON_ID, componentId, handleSubmit, [handleSubmit]);
```
Comment (mickmister) [inline app/screens/apps_form/apps_form_component.tsx:371]: The `SUBMIT` button at the top right of the screen doesn't run the callback for some reason, so I've assigned the close button's callback to the submit handler for now

File: app/screens/post_options/options/app_bindings_post_option.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                    defaultMessage: 'App response type not supported. Response type: {type}.',
+                }, {
+                    type: callResp.type,
+                });
+                postEphemeralCallResponseForPost(serverUrl, callResp, errorMessage, post);
+            }
+        }
+    }, [bindings, post, serverUrl, teamId, intl]);
+
+    const icon = ''; // TODO: Support icons
```
Comment (mickmister) [inline app/screens/post_options/options/app_bindings_post_option.tsx:84]: We need to support HTTP urls for these icons. Here's where the icon string is currently being used:

https://github.com/mattermost/mattermost-mobile/blob/28765a9aaffab435e580524351d0662276be3fa3/app/components/option_item/index.tsx#L238-L242

We may be able to support this as it was in mobile v1, using `FastImage`:

https://github.com/mattermost/mattermost-mobile/blob/df7945318e0c49a1a0047871ab10e05525914caf/app/screens/post_options/post_option.tsx#L58-L67

File: app/screens/post_options/options/app_bindings_post_option.tsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    const options = bindings.map((binding) => (
```
Comment (larkox) [inline app/screens/post_options/options/app_bindings_post_option.tsx:86]: This can be arbitrarily big, so we may consider a different approach here. Depends on what is decided regarding the behaviour of the post options bottom sheet, since may imply adding a flatlist at a higher level.

**Review by larkox — COMMENTED** [review]
In general LGTM. There is still the problem on how we plan to manage the post options.

File: app/screens/apps_form/apps_form_component.tsx [hunk 4]
```diff
@@ -153,24 +144,58 @@ function AppsFormComponent({
...
+    const updateErrors = useCallback((elements: DialogElement[], fieldErrors?: {[x: string]: string}, formError?: string): boolean => {
+        let hasErrors = false;
+        let hasHeaderError = false;
+        if (formError) {
+            hasErrors = true;
+            hasHeaderError = true;
+            setError(formError);
+        }
+
+        if (fieldErrors && Object.keys(fieldErrors).length > 0) {
```
Comment (javaguirre) [inline app/screens/apps_form/apps_form_component.tsx:175]: This would be nice in a validation function, but I see this code was this way before, so I wouldn't do it now.

File: app/screens/apps_form/apps_form_component.tsx [hunk 5]
```diff
@@ -269,21 +264,21 @@ function AppsFormComponent({
             }
         });
 
+        setError('');
```
Comment (javaguirre) [inline app/screens/apps_form/apps_form_component.tsx:267]: why is this needed?

**Review by javaguirre — COMMENTED** [review]
Just some small details that would be nice, otherwise LGTM!

File: app/screens/apps_form/apps_form_component.tsx (hunk 5 above)
Comment (javaguirre) [reply app/screens/apps_form/apps_form_component.tsx:267]: yep, I see, my mistake, there is a `setError` and a `setErrors`, I thought we were modifying the same state variable.

File: app/screens/post_options/post_options.tsx [hunk 6]
```diff
@@ -3,7 +3,9 @@
 
 import {useManagedConfig} from '@mattermost/react-native-emm';
 import React from 'react';
+import {ScrollView} from 'react-native';
```
Comment (enahum) [inline app/screens/post_options/post_options.tsx:6]: From gesture handler, also how many items max could be added? If no limit ScrollView is not the way to go as is not opimized for perf

File: app/screens/post_options/post_options.tsx [hunk 7]
```diff
@@ -69,7 +73,7 @@ const PostOptions = ({
 
     const renderContent = () => {
         return (
-            <>
+            <ScrollView>
```
Comment (enahum) [inline app/screens/post_options/post_options.tsx:76]: Should have scrolling disabled if jo bindings are present

Comment (enahum) [reply app/screens/post_options/post_options.tsx:76]: Maybe I missed it, but I'm not seeing where the snap point of the bottom sheet is added when there are bindings so that it (bottom sheet) can be expanded up to 90% of the screen height, well I'm not sure if that is part of the design or not, but in my mind it made sense

Comment (mickmister) [reply app/screens/post_options/post_options.tsx]: > Maybe I missed it, but I'm not seeing where the snap point of the bottom sheet is added when there are bindings so that it (bottom sheet) can be expanded up to 90% of the screen height, well I'm not sure if that is part of the design or not, but in my mind it made sense

@enahum I'm not familiar with how to do this. Can you point me to a similar solution in another scenario?

File: app/screens/post_options/post_options.tsx (hunk 6 above)
Comment (mickmister) [reply app/screens/post_options/post_options.tsx]: On the order of O(n) apps installed, probably less then n. This is a temporary solution until we have a separate list for actions

File: app/screens/post_options/index.ts [hunk 8]
```diff
@@ -57,7 +57,7 @@ const observeCanEditPost = (database: Database, isOwner: boolean, post: PostMode
 const withPost = withObservables([], ({post, database}: {post: Post | PostModel} & WithDatabaseArgs) => {
     let id: string | undefined;
     let combinedPost: Observable<Post | PostModel | undefined> = of$(undefined);
-    if (post.type === Post.POST_TYPES.COMBINED_USER_ACTIVITY && post.props?.system_post_ids) {
+    if (post?.type === Post.POST_TYPES.COMBINED_USER_ACTIVITY && post.props?.system_post_ids) {
```
Comment (mickmister) [inline app/screens/post_options/index.ts:60]: Not sure why, but the app kept crashing due to `post` sometimes being undefined on app reload

File: app/screens/post_options/post_options.tsx (hunk 7 above)
Comment (enahum) [reply app/screens/post_options/post_options.tsx]: @mickmister basically here https://github.com/mattermost/mattermost-mobile/pull/6735/files#diff-54601045814137800ace02ee7d0134d8cfb24c938ac979c85bde85ae9399b0faR149 you have the snap points of the bottom sheet (which is an array representing the point in the Y axis, consider that as the height of the bottom sheet).

Snap points are set from top to bottom, for the post options you'll see that is the height of each option multiplied by the amount of options, and the second value in the array is the value `10` meaning that the second snap point is a height of 10 from the bottom of the screen, which will cause the bottom sheet to close. Another type of value that the bottom sheet accepts as a snap point is a percentage, like `'90%'` (notice percentages are strings)

Also as you can see as one of the props there is `initialSnapIndex` and this one is hardcoded at the moment to 0, which is the position of the value in the snapPoints array, so here we are saying when showing, start at the position (amount of option items * height of option items)

What I expect is that, when bindings are set, the snapPoints array contains the values `['90%', ((snapPoints + additionalSnapPoints) * ITEM_HEIGHT), 10]` and the initialSnapIndex to be set to 1 so that the bottom sheet continues to open as is today but that can be dragged almost all the way to the top of the screen when there are bindings set.

Comment (mickmister) [reply app/screens/post_options/post_options.tsx]: @enahum For the purpose of 1.0, I think the current behavior of the PR is fine for the current state of Apps. It mainly needs to be functional as we continue to develop the Apps. We can essentially assume that there will be no Apps installed at the moment.

However if there is something wrong here and these changes you suggest are required for this PR, is someone from your team able to implement this? I'm not very familiar with RN or the details of what you're suggesting

File: app/screens/apps_form/apps_form_component.tsx [hunk 9]
```diff
@@ -126,18 +126,9 @@ function AppsFormComponent({
...
-        }
-    }, [serverUrl, componentId, submitting]);
-
-    useNavButtonPressed(CLOSE_BUTTON_ID, componentId, close, [close]);
-    useNavButtonPressed(SUBMIT_BUTTON_ID, componentId, onHandleSubmit, [onHandleSubmit]);
-
     useDidUpdate(() => {
         dispatchValues({elements: form.fields});
-    }, [form]);
+    }, [form, dispatchValues]);
```
Comment (larkox) [inline app/screens/apps_form/apps_form_component.tsx:131]: There is no need to depend on `dispatchValues` since the dispatch function should not change between renders.

File: app/screens/post_options/index.ts (hunk 8 above)
Comment (larkox) [reply app/screens/post_options/index.ts:60]: This is weird. What do you mean exactly by reload?

Comment (mickmister) [reply app/screens/post_options/index.ts:60]: @larkox This is during development, and I believe it can occur when the app is refreshed from the JS server. I will try and pinpoint this issue

File: app/screens/post_options/post_options.tsx (hunk 7 above)
Comment (mickmister) [reply app/screens/post_options/post_options.tsx]: @enahum @larkox What are your thoughts on deferring this item? @javaguirre Do you think you would be able to help implement this?

Comment (larkox) [reply app/screens/post_options/post_options.tsx]: I am ok deferring this item to the next release based in the assumption we will be having 2 or 3 post menu bindings for now.

File: app/hooks/apps.ts [hunk 10]
```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {useCallback} from 'react';
+import {useIntl} from 'react-intl';
+
+import {handleBindingClick} from '@actions/remote/apps';
+import {handleGotoLocation} from '@actions/remote/command';
+import {AppCallResponseTypes} from '@app/constants/apps';
```
Comment (larkox) [inline app/hooks/apps.ts:9]: In general, we try to avoid `@app/` imports. It would be great if you could change these imports like `@app/XXX...` to `@XXX...`.

File: app/screens/apps_form/apps_form_component.tsx [hunk 11]
```diff
@@ -153,24 +144,58 @@ function AppsFormComponent({
             undefined,
             intl.formatMessage({id: 'interactive_dialog.submit', defaultMessage: 'Submit'}),
         );
-        base.enabled = submitting;
+        base.enabled = !submitting;
         base.showAsAction = 'always';
         base.color = theme.sidebarHeaderTextColor;
         return base;
-    }, [theme.sidebarHeaderTextColor, intl, Boolean(submitButtons)]);
+    }, [theme, Boolean(submitButtons), submitting, intl]);
```
Comment (larkox) [inline app/screens/apps_form/apps_form_component.tsx:151]: Here it is interesting to have the `sidebarHeaderTextColor` on the theme. That property is a base type (a string), so when the theme changes, even if it is a new string, a comparison can be made, and not re-run this if it is the same string.

**Review by larkox — APPROVED** [review]
LGTM, just two nits.

**Review by javaguirre — APPROVED** [review]
LGTM! Great work @mickmister ! 💯

File: app/screens/post_options/post_options.tsx (hunk 7 above)
Comment (mickmister) [reply app/screens/post_options/post_options.tsx]: @enahum I changed the code to make room for at most 4 bindings,. I tried the `90%` snap point, but that ends up with a lot of extra whitespace at the bottom of the sheet:

image

image

I'm thinking we should get a UX opinion on this. Who would be the go-to person to discuss this with?

Comment (enahum) [reply app/screens/post_options/post_options.tsx]: For UX that would be @matthewbirtch 

But the approach about modifying the snap point like this `snapPoints += Math.min(bindings.length, 4);` is not what I had in mind, but could work if instead of going full height, we just give it enough so that the user understanda that they can scroll for more items. Makes sense?

**Comment by DHaussermann** [comment]
@mickmister I have done a round of testing post menu bindings as well as channel header bindings seem to be working. 🎉 

Couple issues I see. 
1. Forms don't always work. Sometimes they show up as blank on mobile. Ex: run `/test form full` in the input box. The form is blank and the `X` and submit button are not functional. 
image

2. Forms with embedded form are now showing on mobile. However the App crashes when I try to use the single selects to pick data  
image
(Android crash)

Please reach out if you need any help reproducing these or there is anything I can do to speed up the investigation.

File: app/screens/post_options/post_options.tsx (hunk 7 above)
Comment (mickmister) [reply app/screens/post_options/post_options.tsx]: @enahum Here are two different options:

Partially showing the first binding. I'm thinking this one is best, since the binding options are not going to be as used as the other options.

image

Showing the first binding in whole, and partially showing the second binding

image


-----

In either case, here's what it looks like expanded:

image


Here's what it looks like with 8 options:

image

14 options:

image

**Comment by mickmister** [comment]
> 1. Forms don't always work. Sometimes they show up as blank on mobile. Ex: run `/test form full` in the input box. The form is blank and the `X` and submit button are not functional.
>    image

@DHaussermann This is fixed in the latest commit. Please try the latest build.

> 2. Forms with embedded form are now showing on mobile. However the App crashes when I try to use the single selects to pick data
>    image
>    (Android crash)

This is related to the `IntegrationsSelector` component currently in progress https://github.com/mattermost/mattermost-mobile/pull/6716. Any selects from the following components will not work until it is merged:

- interactive messages
- interactive dialogs
- embedded forms
- app form modals

Also, please make sure the command parser is working correctly. I fixed a few things and it now seems generally stable.

**Comment by DHaussermann** [comment]
@mickmister let's follow-up on this maybe. I don't see a difference in the blank form.

File: app/components/post_list/post/body/content/embedded_bindings/menu_binding/index.tsx [hunk 12]
```diff
@@ -3,18 +3,14 @@
...
 import withObservables from '@nozbe/with-observables';
-import React, {useCallback, useState} from 'react';
-import {useIntl} from 'react-intl';
+import React, {useCallback, useMemo, useState} from 'react';
 import {map} from 'rxjs/operators';
 
-import {handleBindingClick, postEphemeralCallResponseForPost} from '@actions/remote/apps';
-import {handleGotoLocation} from '@actions/remote/command';
+import {postEphemeralCallResponseForPost} from '@actions/remote/apps';
+import {useAppBinding} from '@app/hooks/apps';
```
Comment (enahum) [inline app/components/post_list/post/body/content/embedded_bindings/menu_binding/index.tsx:10]:
```suggestion
import {useAppBinding} from '@hooks/apps';
```

File: app/components/post_list/post/body/content/message_attachments/action_menu/index.tsx [hunk 13]
```diff
@@ -25,21 +25,22 @@ const ActionMenu = ({dataSource, defaultOption, disabled, id, name, options, pos
     }
     const [selected, setSelected] = useState(isSelected?.value);
 
-    const handleSelect = useCallback(async (selectedItem: string | string[]) => {
-        if (Array.isArray(selectedItem)) { // Since AutocompleteSelector is not multiselect, we are sure we only receive a string
+    const handleSelect = useCallback(async (selectedItem: DialogOption | DialogOption[] | undefined) => {
```
Comment (enahum) [inline app/components/post_list/post/body/content/message_attachments/action_menu/index.tsx:28]: `DialogOption | DialogOption[] | undefined` there was a type defined somewhere else for this.

Perhaps we should define all those types in the types folder.. it can be done in a follow up PR

File: app/screens/channel_info/app_bindings/index.tsx [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+        />
+    );
+};
+
+type OwnProps = {
+    channelId: string;
+    serverUrl: string;
+}
+
+const enhanced = withObservables([], (ownProps: WithDatabaseArgs & OwnProps) => {
```
Comment (enahum) [inline app/screens/channel_info/app_bindings/index.tsx:82]: not now, but this file should be split.

File: app/screens/post_options/options/app_bindings_post_option.tsx [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    );
+};
+
+type OwnProps = {
+    post: PostModel;
+    bindings: AppBinding[];
+}
+
+const withTeamId = withObservables(['post'], ({post}: OwnProps) => ({
+    teamId: post.channel.observe().pipe(map((channel: ChannelModel) => channel.teamId)),
```
Comment (enahum) [inline app/screens/post_options/options/app_bindings_post_option.tsx:90]: we should split this to it's own file for readability (0/5 on doing it now or in a different PR)

Now here is the important bit, if the channel for some reason has not been loaded this will cause a crash.
To avoid it, you can change it with the following code

```suggestion
    teamId: post.channelId ? observeChannel(database, post.channelId).pipe(
        switchMap((c) => c.teamId),
        distinctUntilChanged(),
    ) : of$('')
```

File: app/screens/post_options/post_options.tsx [hunk 16]
```diff
@@ -73,64 +81,85 @@ const PostOptions = ({
...
         />
     );
 };
 
-export default PostOptions;
+type OwnProps = {
+    serverUrl: string;
+}
+
+const withBindings = withObservables([], (ownProps: OwnProps) => ({
```
Comment (enahum) [inline app/screens/post_options/post_options.tsx:161]: and this should be moved to the index file 5/5

File: app/screens/post_options/post_options.tsx [hunk 17]
```diff
@@ -73,64 +81,85 @@ const PostOptions = ({
...
+    bindings: AppsManager.observeBindings(ownProps.serverUrl, AppBindingLocations.POST_MENU_ITEM),
+}));
+
+export default React.memo(withServerUrl(withBindings(PostOptions)));
```
Comment (enahum) [inline app/screens/post_options/post_options.tsx:165]: the serverUrl could be added in `app/components/post_list/post/post.tsx?L#185` then you don't need to add the `withServerUrl` HOC

File: app/screens/post_options/post_options.tsx [hunk 18]
```diff
@@ -73,64 +81,85 @@ const PostOptions = ({
...
 
     return (
         <BottomSheet
             renderContent={renderContent}
             closeButtonId={POST_OPTIONS_BUTTON}
             componentId={Screens.POST_OPTIONS}
-            initialSnapIndex={0}
-            snapPoints={[((snapPoints + additionalSnapPoints) * ITEM_HEIGHT), 10]}
+            initialSnapIndex={1}
+            snapPoints={[upperSnapPoints * ITEM_HEIGHT, lowerSnapPoints * ITEM_HEIGHT, 10]}
```
Comment (enahum) [inline app/screens/post_options/post_options.tsx:151]: I don't see the reason of having three snap points when there are no bindings..

in short, this should change if there are app bindings, if not it should remain the exact way it was before this PR.

**Review by enahum — CHANGES_REQUESTED** [review]

**Comment by enahum** [comment]
Additionally there are conflicts to be resolved

File: app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts [hunk 19]
```diff
@@ -978,12 +978,15 @@ export class AppCommandParser {
                         const userID = (f.value as AppSelectOption).value;
                         let user: UserModel | UserProfile | undefined = await getUserById(this.database, userID);
                         if (!user) {
-                            const res = await fetchUsersByIds(this.serverUrl, [userID]);
+                            const res = await fetchUsersByIds(this.serverUrl, [userID], true);
```
[the remark this answers is not in the document]
Comment (mickmister) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:981]: @enahum If the user id provided by the App does not match an existing user, the mobile app crashes here https://github.com/mattermost/mattermost-mobile/blob/068549a2852bff605cad2097697e26b47fe7503e/app/actions/remote/user.ts#L411. What do you suggest as an alternative?

[the remark this answers is not in the document]
Comment (enahum) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:981]: Just the last argument leave it as the defsult false so that it get's added to the database and then leave the rest as is

[the remark this answers is not in the document]
Comment (mickmister) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:981]: @enahum If it's set to false, the mobile app crashes if the user id is not an existing user on the MM server. There has to be some change here to avoid the crash

[the remark this answers is not in the document]
Comment (enahum) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:981]: I don't follow, why would it not be an existing user in the server? And in that case, then I think the action should be fixed to handle that

[the remark this answers is not in the document]
Comment (mickmister) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:981]: @enahum The App can provide any value for any of the fields. If it sets an invalid user id as the initial value of a user autocomplete field, that's when this occurs. I'll look into how to avoid this but idk the specifics of what's happening behind the scenes with `operator.handleUsers`

File: app/screens/post_options/options/app_bindings_post_option.tsx (hunk 15 above)
Comment (mickmister) [reply app/screens/post_options/options/app_bindings_post_option.tsx]: @enahum I fixed the missing channel issue here. LMK what you think

File: app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts:981]: @enahum This is the warning that shows when an invalid user id is provided:

image

[the remark this answers is not in the document]
Comment (mickmister) [reply app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser.ts]: I have this fixed in the latest commit

**Comment by mickmister** [comment]
@enahum I believe I've addressed everything other than splitting some of the pieces into different files. Please take another look when you have the chance

**Comment by DHaussermann** [comment]
I have done another round of testing on the latest mobile build.
This is going looking good. And there are no issues found with the command parser. 

For embedded posts - the selects are now working and the crash is resolved. However, I don't see the submit action working on mobile the way it works on Webapp. 

On webapp the test app will post a json payload for the submit. On mobile there is no visible effect to tapping submit on an embedded post. 

I'm not sure if this could be an issue with the submit button yje test app is using here or a gap in the implementation of embedded forms.     

When I look at my ngrok tunnel, I **do** see a web request for mobile....
image


Strangely this looks quite different from what I see in the web traffic when I click submit on webapp..
image

@mickmister please advise on what next steps may be needed here. 
cc @javaguirre  if you have any thoughts as I believe this could be related to your PR as well.

**Comment by mickmister** [comment]
> On webapp the test app will post a json payload for the submit. On mobile there is no visible effect to tapping submit on an embedded post.

@DHaussermann I'm not seeing this issue. The submit button works for me on mobile. The JSON response shows in an ephemeral message just as it does for webapp. Let's get together and debug this if you are seeing this issue.

> Strangely this looks quite different from what I see in the web traffic when I click submit on webapp

That traffic seems to be for requests related to the playbooks plugin, and should not be triggered by clicking the submit button.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- For this last round of testing, my confusion was actually caused be how the mobile client handles ephemeral posts in a reply thread with CRT enabled. I will open a separate Jira ticket to see if this may require review.
- The test app does post a response and shows .json as expected.
- With a recent change to the Test-App I was also able to test with a varying the number post bindings to guard against a bug found by @mickmister 
- No know issues remaining 
LGTM!

**Comment by DHaussermann** [comment]
@amyblais I'm not sure of the correct milestone but, this PR is needed for the December release.

File: app/actions/remote/user.ts [hunk 20]
```diff
@@ -407,7 +407,7 @@ export const fetchUsersByIds = async (serverUrl: string, userIds: string[], fetc
             return {users: [], existingUsers};
         }
         const users = await client.getProfilesByIds([...new Set(usersToLoad)]);
-        if (!fetchOnly) {
+        if (!fetchOnly && users.length) {
```
Comment (enahum) [inline app/actions/remote/user.ts:410]: I guess you added this to remove the warning when no users are returned by the client, not needed by more than ok

File: app/components/option_item/index.tsx [hunk 21]
```diff
@@ -125,7 +126,6 @@ const OptionItem = ({
     description,
     destructive,
     icon,
-    iconColor,
```
Comment (enahum) [inline app/components/option_item/index.tsx:128]: If we are no longer using this, then it should be removed from the type definition. But why are we removing it?

File: app/components/option_item/option_icon.tsx [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+
+    const iconAsSource = useMemo(() => {
+        return {uri: icon};
+    }, [icon]);
+
+    if (isValidUrl(icon) && !failedToLoadImage) {
+        return (
+            <FastImage
+                source={iconAsSource}
+                style={imageStyle}
```
Comment (enahum) [inline app/components/option_item/option_icon.tsx:52]: This styles have some colors in them, but to tint the image a tintColor prop is expected, is that being taken into consideration?

File: app/screens/integration_selector/index.tsx [hunk 23]
```diff
@@ -457,19 +456,19 @@ function IntegrationSelector(
...
                 text = {
-                    id: intl.formatMessage({id: 'mobile.integration_selector.loading_users'}),
+                    id: 'mobile.integration_selector.loading_users',
                     defaultMessage: 'Loading Users...',
                 };
                 break;
             case ViewConstants.DATA_SOURCE_CHANNELS:
                 text = {
-                    id: intl.formatMessage({id: 'mobile.integration_selector.loading_channels'}),
+                    id: 'mobile.integration_selector.loading_channels',
```
Comment (enahum) [inline app/screens/integration_selector/index.tsx:465]: these id's needs to be wrapped using the `t` function found in the `@i18n` module.

After that, Make sure to run `npm run i18n-extract` to make sure all strings are present

File: app/screens/post_options/post_options.tsx [hunk 24]
```diff
@@ -73,64 +80,91 @@ const PostOptions = ({
...
 };
 
-export default PostOptions;
+type OwnProps = {
+    serverUrl: string;
+}
+
+const withBindings = withObservables([], (ownProps: OwnProps) => ({
+    bindings: AppsManager.observeBindings(ownProps.serverUrl, AppBindingLocations.POST_MENU_ITEM),
+}));
```
Comment (enahum) [inline app/screens/post_options/post_options.tsx:168]: This still needs to be movef to the index file

**Review by enahum — CHANGES_REQUESTED** [review]
We are almost there, just a few more nits. (Sorry for being so anniying with the review)

Do not forget to run i18n-extract

File: app/components/option_item/index.tsx (hunk 21 above)
Comment (mickmister) [reply app/components/option_item/index.tsx:128]: Thanks for pointing this out. I've added it back by using this in `option_icon.tsx`

File: app/components/option_item/option_icon.tsx (hunk 22 above)
Comment (mickmister) [reply app/components/option_item/option_icon.tsx:52]: @enahum Note that will only render here if this is a URL image. Not a compass image, and also likely not an svg. Should we remove everything "color" about the styling of this `FastImage` element here?

Comment (mickmister) [reply app/components/option_item/option_icon.tsx:52]: I don't think the intended plan is to tint the images at all, in order to keep the branding of the given integration. Do you suggest we tint the color anyway?

**Comment by mickmister** [comment]
Ticket for the remaining issue of radio button styling https://mattermost.atlassian.net/browse/MM-48732. 3/5 should not be a blocker for this release

File: app/screens/post_options/index.ts [hunk 25]
```diff
@@ -6,8 +6,10 @@ import withObservables from '@nozbe/with-observables';
 import {combineLatest, of as of$, Observable} from 'rxjs';
 import {switchMap} from 'rxjs/operators';
 
+import {AppBindingLocations} from '@app/constants/apps';
```
Comment (enahum) [inline app/screens/post_options/index.ts:9]:
```suggestion
import {AppBindingLocations} from '@constants/apps';
```

**Review by enahum — APPROVED** [review]
one final little detail with an import and this should be good to go.

Awesome work @mickmister 🎉

**Comment by mickmister** [comment]
@DHaussermann I'm thinking this is good to merge. The changes since your approval are small in scope, and I've asserted those pieces work correctly. Thoughts on merging this?
